File: pesquisa_operacional/DilemaDoPrisioneiro.java
import java.util.Scanner;
import java.util.Random;

public class DilemaDoPrisioneiro {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Random random = new Random();

        System.out.println("Bem-vindo ao Dilema do Prisioneiro em Java!");

        // Loop principal para permitir múltiplas rodadas
        while (true) {
            System.out.println("\n--- Nova Rodada ---");

            // Escolha do Jogador 1 (Humano)
            String escolhaJogador1 = obterEscolhaHumana(scanner, "Jogador 1");

            // Escolha do Jogador 2 (Computador - Aleatório)
            String escolhaJogador2 = obterEscolhaComputadorAleatorio(random);
            System.out.println("Jogador 2 escolheu: " + (escolhaJogador2.equals("colaborar") ? "colaborar (c)" : "trair (t)"));

            // Determina as sentenças com base nas escolhas
            determinarSentencas(escolhaJogador1, escolhaJogador2, "Jogador 1", "Jogador 2");

            // Pergunta se deseja jogar novamente
            if (!desejaJogarNovamente(scanner)) {
                break;
            }
        }

        System.out.println("\nObrigado por jogar!");
        scanner.close();
    }

    /**
     * Obtém a escolha do jogador humano (colaborar ou trair) através da entrada do scanner.
     *
     * @param scanner O objeto Scanner para leitura da entrada do usuário.
     * @param nomeJogador O nome do jogador.
     * @return A escolha do jogador ("colaborar" ou "trair").
     */
    public static String obterEscolhaHumana(Scanner scanner, String nomeJogador) {
        String escolha;
        while (true) {
            System.out.println(nomeJogador + ", deseja colaborar (c) ou trair (t)?");
            String entrada = scanner.next().toLowerCase();
            if (entrada.equals("c")) {
                escolha = "colaborar";
                break;
            } else if (entrada.equals("t")) {
                escolha = "trair";
                break;
            } else {
                System.out.println("Escolha inválida. Digite 'c' para colaborar ou 't' para trair.");
            }
        }
        return escolha;
    }

    /**
     * Obtém a escolha do computador de forma aleatória.
     *
     * @param random O objeto Random para gerar a escolha aleatória.
     * @return A escolha do computador ("colaborar" ou "trair").
     */
    public static String obterEscolhaComputadorAleatorio(Random random) {
        return random.nextBoolean() ? "colaborar" : "trair";
    }

    /**
     * Determina e exibe as sentenças para ambos os jogadores com base em suas escolhas.
     *
     * @param escolhaJogador1 A escolha do Jogador 1.
     * @param escolhaJogador2 A escolha do Jogador 2.
     * @param nomeJogador1 O nome do Jogador 1.
     * @param nomeJogador2 O nome do Jogador 2.
     */
    public static void determinarSentencas(String escolhaJogador1, String escolhaJogador2, String nomeJogador1, String nomeJogador2) {
        System.out.println("\n--- Resultados da Rodada ---");
        System.out.println(nomeJogador1 + " escolheu: " + (escolhaJogador1.equals("colaborar") ? "colaborar (c)" : "trair (t)"));
        System.out.println(nomeJogador2 + " escolheu: " + (escolhaJogador2.equals("colaborar") ? "colaborar (c)" : "trair (t)"));

        if (escolhaJogador1.equals("colaborar") && escolhaJogador2.equals("colaborar")) {
            System.out.println("Ambos colaboraram. Ambos recebem 2 anos de prisão.");
        } else if (escolhaJogador1.equals("colaborar") && escolhaJogador2.equals("trair")) {
            System.out.println(nomeJogador1 + " colaborou e " + nomeJogador2 + " traiu. " + nomeJogador1 + " recebe 10 anos e " + nomeJogador2 + " sai livre.");
        } else if (escolhaJogador1.equals("trair") && escolhaJogador2.equals("colaborar")) {
            System.out.println(nomeJogador1 + " traiu e " + nomeJogador2 + " colaborou. " + nomeJogador2 + " recebe 10 anos e " + nomeJogador1 + " sai livre.");
        } else { // Ambos traíram
            System.out.println("Ambos traíram. Ambos recebem 5 anos de prisão.");
        }
    }

    /**
     * Pergunta ao jogador se ele deseja jogar novamente.
     *
     * @param scanner O objeto Scanner para leitura da entrada do usuário.
     * @return true se o jogador deseja jogar novamente, false caso contrário.
     */
    public static boolean desejaJogarNovamente(Scanner scanner) {
        System.out.println("\nDeseja jogar novamente? (s/n)");
        String resposta = scanner.next().toLowerCase();
        return resposta.equals("s");
    }
}

File: pesquisa_operacional/DilemadoPrisioneiro.py
import random

def dilema_do_prisioneiro(jogador1, jogador2):
    """
    Simula uma rodada do Dilema do Prisioneiro entre dois jogadores.

    Args:
        jogador1: Uma função que representa a estratégia do jogador 1.
                  Deve retornar 'colaborar' ou 'trair'.
        jogador2: Uma função que representa a estratégia do jogador 2.
                  Deve retornar 'colaborar' ou 'trair'.

    Returns:
        Um dicionário contendo as escolhas de cada jogador e as sentenças resultantes.
    """
    # Obtém a escolha do jogador 1 chamando a função de sua estratégia
    escolha1 = jogador1()
    # Obtém a escolha do jogador 2 chamando a função de sua estratégia
    escolha2 = jogador2()

    # Cria um dicionário para armazenar as escolhas dos jogadores
    resultados = {
        'jogador1': escolha1,
        'jogador2': escolha2
    }

    # Determina as sentenças com base nas escolhas de ambos os jogadores
    if escolha1 == 'colaborar' and escolha2 == 'colaborar':
        resultados['sentenca_jogador1'] = 2  # Ambos colaboram: 2 anos para cada
        resultados['sentenca_jogador2'] = 2
        print("\nAmbos colaboraram. Ambos recebem 2 anos de prisão.")
    elif escolha1 == 'colaborar' and escolha2 == 'trair':
        resultados['sentenca_jogador1'] = 10 # Jogador 1 colabora, Jogador 2 trai: 10 anos para Jogador 1, 0 para Jogador 2
        resultados['sentenca_jogador2'] = 0
        print(f"\nJogador 1 colaborou e Jogador 2 traiu. Jogador 1 recebe 10 anos e Jogador 2 sai livre.")
    elif escolha1 == 'trair' and escolha2 == 'colaborar':
        resultados['sentenca_jogador1'] = 0  # Jogador 1 trai, Jogador 2 colabora: 0 anos para Jogador 1, 10 para Jogador 2
        resultados['sentenca_jogador2'] = 10
        print(f"\nJogador 1 traiu e Jogador 2 colaborou. Jogador 2 recebe 10 anos e Jogador 1 sai livre.")
    else:  # escolha1 == 'trair' and escolha2 == 'trair'
        resultados['sentenca_jogador1'] = 5  # Ambos traem: 5 anos para cada
        resultados['sentenca_jogador2'] = 5
        print("\nAmbos traíram. Ambos recebem 5 anos de prisão.")

    # Retorna um dicionário contendo as escolhas e as sentenças
    return resultados

def jogador_humano():
    """
    Permite que um humano faça sua escolha através da entrada do teclado.
    """
    while True:
        escolha = input("Você deseja colaborar (c) ou trair (t)? ").lower()
        # Verifica se a entrada é 'c' ou 't'
        if escolha in ['c', 't']:
            return 'colaborar' if escolha == 'c' else 'trair'
        else:
            print("Escolha inválida. Digite 'c' para colaborar ou 't' para trair.")

def jogador_aleatorio():
    """
    Faz uma escolha aleatória entre colaborar e trair.
    """
    return random.choice(['colaborar', 'trair'])

def jogador_sempre_colabora():
    """
    Implementa uma estratégia onde o jogador sempre escolhe colaborar.
    """
    return 'colaborar'

def jogador_sempre_trai():
    """
    Implementa uma estratégia onde o jogador sempre escolhe trair.
    """
    return 'trair'

# Bloco principal de execução do script
if __name__ == "__main__":
    print("Bem-vindo ao Dilema do Prisioneiro!")

    # Loop para permitir que o jogo seja jogado várias vezes
    while True:
        print("\nOpções de jogadores:")
        print("1: Humano")
        print("2: Aleatório")
        print("3: Sempre Colabora")
        print("4: Sempre Trai")

        try:
            # Solicita ao usuário que escolha a estratégia para o Jogador 1
            jogador1_opcao = int(input("Escolha o Jogador 1: "))
            # Solicita ao usuário que escolha a estratégia para o Jogador 2
            jogador2_opcao = int(input("Escolha o Jogador 2: "))

            # Cria um dicionário para mapear as opções numéricas às funções de estratégia
            jogadores = {
                1: jogador_humano,
                2: jogador_aleatorio,
                3: jogador_sempre_colabora,
                4: jogador_sempre_trai
            }

            # Verifica se as opções escolhidas são válidas
            if jogador1_opcao in jogadores and jogador2_opcao in jogadores:
                # Obtém as funções de estratégia com base nas opções
                jogador1 = jogadores[jogador1_opcao]
                jogador2 = jogadores[jogador2_opcao]
                # Simula uma rodada do dilema do prisioneiro
                resultados = dilema_do_prisioneiro(jogador1, jogador2)
                # Imprime os resultados da rodada
                print("\n--- Resultados da Rodada ---")
                print(f"Jogador 1 escolheu: {resultados['jogador1']}")
                print(f"Jogador 2 escolheu: {resultados['jogador2']}")
                print(f"Sentença Jogador 1: {resultados['sentenca_jogador1']} anos")
                print(f"Sentença Jogador 2: {resultados['sentenca_jogador2']} anos")
            else:
                print("Opção de jogador inválida.")

            # Pergunta ao usuário se deseja jogar novamente
            jogar_novamente = input("\nDeseja jogar novamente? (s/n): ").lower()
            # Se a resposta não for 's', o loop é interrompido
            if jogar_novamente != 's':
                break

        # Trata a exceção caso o usuário digite algo que não seja um número
        except ValueError:
            print("Por favor, digite um número para escolher o jogador.")

    print("\nObrigado por jogar!")
